File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start = 0
        count = {}
        max_length = 0
        
        for i in range(len(s)):
            if s[i] in count:
                start = max(start, count[s[i]] + 1)
            count[s[i]] = i
            max_length = max(max_length, i - start + 1)
        return max_length

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-3.cs
public class Solution {
    public int LargestRectangleArea(int[] heights) {
        Stack<int> stack = new Stack<int>();
        int maxArea = 0;
        for (int i =0; i <= heights.Length; i++) {
            int height = i == heights.Length ? 0 : heights[i];
            while(stack.Count > 0 && height < heights[stack.Peek()]) {
                int h = heights[stack.Pop()];
                int w;
                if (stack.Count == 0) {
                    w = i;
                }else {
                    w = i - stack.Peek() - 1;
                }
                maxArea = Math.Max(maxArea, w * h);
            }
            stack.Push(i);
        }
        return maxArea;
    }
}
